pkg/kubelet/images/image_pull_manager.go

```diff
@@ -0,0 +1,663 @@
+	return &NamedLockSet{
+		globalLock: sync.Mutex{},
+		locks:      map[string]*sync.Mutex{},
+	}
+}
+
+func (n *NamedLockSet) Lock(name string) {
+	func() {
+		n.globalLock.Lock()
+		defer n.globalLock.Unlock()
+		if _, ok := n.locks[name]; !ok {
+			n.locks[name] = &sync.Mutex{}
+		}
+	}()
+	// This call cannot be guarded by the global lock as it would block the access
+	// to the other locks
+	n.locks[name].Lock()
+}
+
+// Unlock unlocks the named lock. Can only be called after a previous Lock() call
+// for the same named lock.
+func (n *NamedLockSet) Unlock(name string) {
+	// cannot be locked by the global lock as it would deadlock once GlobalLock() gets activated
+	if _, ok := n.locks[name]; ok {
+		n.locks[name].Unlock()
+	}
+}
+
+// GlobalLock first locks access to the named locks and then locks all of the
+// set locks
+func (n *NamedLockSet) GlobalLock() {
+	n.globalLock.Lock()
+	for _, l := range n.locks {
+		l.Lock()
+	}
+}
+
+// GlobalUnlock should only be called after GlobalLock(). It unlocks all the locks
+// of the set and then it also unlocks the global lock gating access to the set locks.
+func (n *NamedLockSet) GlobalUnlock() {
+	for _, l := range n.locks {
+		l.Unlock()
+	}
+	n.globalLock.Unlock()
+}
+
+// FileBasedImagePullManager is an implementation of the ImagePullManager. It
+// tracks images pulled by the kubelet by creating records about ongoing and
+// successful pulls.
+// It tracks the credentials used with each successful pull in order to be able
+// to distinguish tenants requesting access to an image that exists on the kubelet's
+// node.
+type FileBasedImagePullManager struct {
+	pullingDir string
+	pulledDir  string
+
+	imagePullPolicy ImagePullPolicyEnforcer
+
+	imageService kubecontainer.ImageService
+
+	intentAccessors *NamedLockSet  // image -> sync.Mutex
+	intentCounters  map[string]int // image -> number of current in-flight pulls
+
+	pulledAccessors *NamedLockSet // imageRef -> sync.Mutex
+
+	encoder runtime.Encoder
+	decoder runtime.Decoder
+}
+
+func NewFileBasedImagePullManager(ctx context.Context, kubeletDir string, imagePullPolicy ImagePullPolicyEnforcer, imageService kubecontainer.ImageService) (*FileBasedImagePullManager, error) {
+	kubeletConfigEncoder, kubeletConfigDecoder, err := createKubeletConfigSchemeEncoderDecoder()
+	if err != nil {
+		return nil, err
+	}
+
+	m := &FileBasedImagePullManager{
+		pullingDir: filepath.Join(kubeletDir, "image_manager", "pulling"),
+		pulledDir:  filepath.Join(kubeletDir, "image_manager", "pulled"),
+
+		imagePullPolicy: imagePullPolicy,
+
+		imageService: imageService,
+
+		intentAccessors: NewNamedLockSet(),
+		intentCounters:  make(map[string]int),
+
+		pulledAccessors: NewNamedLockSet(),
+
+		encoder: kubeletConfigEncoder,
+		decoder: kubeletConfigDecoder,
+	}
+
+	if err := os.MkdirAll(m.pullingDir, 0700); err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(m.pulledDir, 0700); err != nil {
+		return nil, err
+	}
+
+	m.startupHousekeeping(ctx)
```

Comment:
nit: "startup" sounds like this is spawning a background goroutine ... suggest a name like `initialize()` if this is doing blocking things that will complete before this returns